sdlib/d/gc/tcache.d

```diff
@@ -244,6 +274,74 @@ public:
 	}
 
 private:
+	capacityInfo getAllocInfo(PageDescriptor pd, void* ptr) {
+		if (pd.extent.isLarge()) {
+			return capacityInfo(pd.extent.address, pd.extent.size,
+			                    pd.extent.usedCapacity);
+		}
+
+		// Slab alloc:
+		auto sg = slabAllocGeometry(ptr, pd, false);
+
+		// If label flag is 0, or this size class does not support labels,
+		// then the alloc is reported to be fully used:
+		if (!sg.e.hasLabel(sg.index)) {
+			return capacityInfo(sg.address, sg.size, sg.size);
+		}
+
+		// Decode label, found in the final byte (or two bytes) of the alloc:
+		auto body = (cast(ubyte*) sg.address);
+		ubyte lo = body[sg.size - 1];
+		ushort freeSize = lo >>> 1;
+		if (lo & 1 != 0) {
+			ushort hi = body[sg.size - 2];
+			freeSize |= hi << 7;
+		}
```

Comment:
You should probably have a scheme that loads a short and make the irrelevant parts. For instance:
```d
auto data = *(cast(ushort*) ptr);
auto mask = 0x7f | -(data & 0x01) ;
auto freeBytes = (data >> 1) & mask;
```